lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -374,31 +373,28 @@ def write(chunk)
             if record.key?('kubernetes')
               handle_container_metadata(record, entry)
             end
-            # Save the timestamp if available
-            timestamp = record.key?('time') ? record['time'] : nil
-            # If the log from the user container is json, we want to export it
-            # as a structured log. Now that we've pulled out all the
-            # container-specific metadata from the record, we can replace the
-            # record with the json that the user logged.
-            # To save CPU in the common case of unstructured logs, only check if
-            # the contents are parsable as json for the first entry of each
-            # batch.
-            if is_container_json.nil? && record.key?('log')
-              record_json = parse_json_or_nil(record['log'])
-              if record_json.nil?
-                is_container_json = false
-              else
-                record = record_json
-                is_container_json = true
-              end
-            elsif is_container_json && record.key?('log')
-              record_json = parse_json_or_nil(record['log'])
-              record = record_json unless record_json.nil?
-            end
-            # Restore timestamp if necessary
-            unless record.key?('time') || timestamp.nil?
-              record['time'] = timestamp
-            end
+          end
+
+          # Save the timestamp if available, then clear it out to allow for
+          # determining whether we should parse the log or message field.
+          timestamp = record.key?('time') ? record['time'] : nil
+          record.delete('time')
+          # If the log is json, we want to export it as a structured log unless
+          # there is additional metadata that would be lost.
+          is_json = false
+          if record.length == 1 && record.key?('log')
+            record_json = parse_json_or_nil(record['log'])
```

Comment:
My hope was to handle JSON-shaped text not just in the container case. Specifically, in the case of App Engine Flex, because the user code runs in a docker container, logs are text-shaped and currently there isn't any way for users to modify the fluentd configuration to specify the "format json" option.